Fix unavailable property for wemo switch

balloob opened the pull request:
## Description:
This will mark Wemo switches as unavailable if they haven't been able to update. 

**Related issue (if applicable):** #12935 

## Example entry for `configuration.yaml` (if applicable):
```yaml
wemo:
```

## Checklist:
  - [x] The code change is tested and works locally.
  - [x] Local tests pass with `tox`. **Your PR cannot be merged unless tests pass**

[ex-requir]: https://github.com/home-assistant/home-assistant/blob/dev/homeassistant/components/keyboard.py#L14
[ex-import]: https://github.com/home-assistant/home-assistant/blob/dev/homeassistant/components/keyboard.py#L54

balloob committed:
Fix unavailable property for wemo switch

balloob commented:
@arsaboo I've tested this PR locally and it works. Could you also test this?

Testing steps:
  - Start HASS and have it find your Wemo switch
  - Disconnect Wemo switch
  - HASS should mark Wemo switch as unavailable
  - Plug Wemo switch back in
  - HASS should make Wemo switch available again

MartinHjelmare commented on homeassistant/components/switch/wemo.py:
```diff
@@ -199,9 +201,33 @@ def turn_off(self, **kwargs):
         self.wemo.off()
         self.schedule_update_ha_state()
```
This should probably be removed, since we have callbacks for state change called from pywemo.

MartinHjelmare commented on homeassistant/components/switch/wemo.py:
```diff
@@ -199,9 +201,33 @@ def turn_off(self, **kwargs):
+        Wemo switch is unreachable. If update goes through, it will be made
+        available again.
+        """
+        if self._update_lock.locked():
```
This doesn't protect `_update` from being called excessively both from callback and poll.

MartinHjelmare reviewed (commented):
Adding polling also adds calls to `async_update` after every service call. The lock as currently placed will not protect from unnecessary calls to `_update`.

Side note, this module would benefit from a clean up according to our current patterns and practices. Eg where/when callbacks should be registered.

arsaboo commented:
Tested this and it works as expected 👍

After taking the Wemo Insight out, it shows `unavailable` in about 5-7 seconds. After I plug it back in, the switch becomes available as soon as the network is established (typically about 30 seconds or so).

balloob committed:
Have subscriptions respect the lock

MartinHjelmare commented on homeassistant/components/switch/wemo.py:
```diff
@@ -65,17 +65,28 @@ def __init__(self, device):
         wemo = get_component('wemo')
         wemo.SUBSCRIPTION_REGISTRY.register(self.wemo)
-        wemo.SUBSCRIPTION_REGISTRY.on(self.wemo, None, self._update_callback)
+        wemo.SUBSCRIPTION_REGISTRY.on(
```
Move this to `async_added_to_hass`.

MartinHjelmare commented on homeassistant/components/switch/wemo.py:
```diff
@@ -54,29 +58,44 @@ def __init__(self, device):
         updated = self.wemo.subscription_update(_type, _params)
-        self._update(force_update=(not updated))
 
         if not hasattr(self, 'hass'):
```
This can be removed after the move of callback registration to `async_added_to_hass`.

balloob commented:
Addressed the first round of comments by @MartinHjelmare . (Not the ones that arrived while I was typing this)

I agree that it needs a big cleanup. This is one of the first integrations that was added to Home Assistant. The fact that a call to update() can block up to 5 minutes is something I added to pywemo back in the day… I wouldn't do that today anymore. However, I want to fix the available thing right now, maybe one day an async rewrite of pywemo is in order.

balloob committed:
Move subscription callback to added to hass section

balloob commented:
Addressed 2nd round of comments 👍

I did not get rid of the global for wemo because I don't have devices to test the light/binary sensor platforms.

MartinHjelmare reviewed (approved):
Looks good!
